src/server/serve.rs

```diff
@@ -54,148 +61,179 @@ pub fn serve(
     port: u32,
     filter_search_domain: &str,
 ) -> Result<(), std::io::Error> {
+    let mut signals = Signals::new([SIGHUP])?;
+
+    let (backend, mut listen_ip_v4, mut listen_ip_v6) = parse_configs(config_path)?;
+
+    // We store the `DNSBackend` in an `ArcSwap` so we can replace it when the configuration is
+    // reloaded.
+    static DNSBACKEND: OnceLock<ArcSwap<DNSBackend>> = OnceLock::new();
+    let backend = DNSBACKEND.get_or_init(|| ArcSwap::from(Arc::new(backend)));
+
+    let mut handles_v4 = HashMap::new();
+    let mut handles_v6 = HashMap::new();
+
     loop {
-        if let Err(er) = core_serve_loop(config_path, port, filter_search_domain) {
-            return Err(std::io::Error::new(
-                std::io::ErrorKind::Other,
-                format!("Server Error {}", er),
-            ));
+        debug!("Successfully parsed config");
+        debug!("Listen v4 ip {:?}", listen_ip_v4);
+        debug!("Listen v6 ip {:?}", listen_ip_v6);
+
+        // kill server if listen_ip's are empty
+        if listen_ip_v4.is_empty() && listen_ip_v6.is_empty() {
+            info!("No configuration found stopping the sever");
+
+            let path = Path::new(config_path).join(AARDVARK_PID_FILE);
+            if let Err(err) = fs::remove_file(path) {
+                error!("failed to remove the pid file: {}", &err);
+                process::exit(1);
+            }
+
+            // Gracefully stop all server threads first.
+            stop_threads(&mut handles_v4, None);
+            stop_threads(&mut handles_v6, None);
+
+            process::exit(0);
+        }
+
+        stop_and_start_threads(
+            port,
+            filter_search_domain,
+            backend,
+            listen_ip_v4,
+            &mut handles_v4,
+        );
+
+        stop_and_start_threads(
+            port,
+            filter_search_domain,
+            backend,
+            listen_ip_v6,
+            &mut handles_v6,
+        );
+
+        // Block until we receive a SIGHUP.
+        loop {
+            if signals.wait().next().is_some() {
+                break;
+            }
         }
+
+        let (new_backend, new_listen_ip_v4, new_listen_ip_v6) = parse_configs(config_path)?;
+
+        // Only replace the backend value if the configuration actually changed.
+        if **backend.load() != new_backend {
```

Comment:
My understanding is that this walks the entire map of names/ips and so on to compare if anything changed right? I guess that could get expensive with many containers, networks....
Also in general we only send SIGHUP when we actually changed something so I would suggest to not check this and always replace

Reply:
Fair enough, done.